Fix flaky navbar interaction on Notification: Actions test

john-legg opened the pull request:
### Context
An attempt to fix the `Notification: Actions` e2e test, which has been [intermittently failing lately](https://github.com/Automattic/wp-calypso/issues/69857). This is part of a previous effort to [improve and un-quarantine the test](https://github.com/Automattic/wp-calypso/pull/69446). After a chat session with Edwin, we identified the problem area (navbar loading) and a few potential solutions:

- Update pageSettled() in navbar-component.ts to use methods more in-line with Playwright's current suggestions.
- Wait for some other object on the navbarComponent to load first (less ideal).
- Add a general wait to the test (least ideal).

john-legg committed:
Fix flaky navbar interaction on Notification: Actions test

john-legg commented:
### Update/Findings

Hi @worldomonation, I don't have too much progress to show, but here are my notes and thought process. 

Slow connection doesn't seem to be the main issue. From debugging trace files, it seems that the majority of the page is rendered in less than 8 seconds (rough average). The navbar (containing the notification icon) loads before the rest of the home page. 

During navbar loading, the notification icon always loads first, while components such as the `Write` and `help` buttons load last. At the first moment that the notification icon is visible, the rest of the home page hasn't started loading yet. After talking with Edwin, I suspect that since the notification icon is visible for a few milliseconds before everything else, the test perceives this as 'stable' or 'loaded' even though many elements are not loaded or interact-able yet.

I spent a lot of time trying to replicate this locally - I even slowed the throttle _way_ down and it didn't cause any of the same behaviors. I also tried to elegantly wait for the navbar to fully load, but I couldn't seem to get it right (I'm not as savvy with Playwright as I thought 😓). Suggestions welcomed!

A less-ideal alternative that Edwin mentioned would be to wait for another navbar element to load before proceeding with any navbar interaction, so that's what I did. I settled on the `help` button since it's visible on both desktop and mobile. Not great, but open to feedback! Here's a debug of the milliseconds this step adds to the test:

```
pw:api waiting for selector "[title="Help"]" to be visible +2ms
  pw:api   navigated to "http://calypso.localhost:3000/" +155ms
  pw:api   navigated to "about:blank" +161ms
  pw:api   selector resolved to visible <button title="Help" class="masterbar__item masterbar__i…>…</button> +12ms
  pw:api <= page.waitForSelector succeeded +91ms
```
Unfortunately, I don't have a way to replicate the original issue (throttling hasn't helped), so I have no idea if it solves the flakiness or not. Any ideas how to test this out?

worldomonation commented:
@john-legg FYI, if the PR is not complete, it should not be set to `Needs Review` status.

worldomonation committed:
packages/calypso-e2e/src/element-helper.ts
- add a generic method to wait for the `/home` page to finish loading as much as possible.

packages/calypso-e2e/src/lib/components/navbar-component.ts
- removed use of `pageSettled` in favor of `ElementHelper.waitForHomeLoad`
- replace use of `waitForSelector` in `openNotificationsPanel` with the Locator API.

worldomonation committed:
packages/calypso-e2e/src/element-helper.ts
- fix incorrect function name

worldomonation committed:
packages/calypso-e2e/src/element-helper.ts
- remove changes

packages/calypso-e2e/src/lib/components/navbar-component.ts
- restrict the `waitForHome` to just the `openNotificationsPanel` method.

worldomonation committed:
packages/calypso-e2e/src/lib/components/navbar-component.ts
- restore wait for `async-load-calypso-blocks-jitm.*/` as additional guard for desktop viewports

worldomonation committed:
packages/calypso-e2e/src/lib/components/navbar-component.ts
- remove unused selector

worldomonation commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -71,20 +60,22 @@ export class NavbarComponent {
 		useKeyboard = false,
 	}: { useKeyboard?: boolean } = {} ): Promise< void > {
-		await this.pageSettled();
+		const promises: [ Promise< unknown > ] = [ this.page.waitForLoadState( 'networkidle' ) ];
+
+		if ( envVariables.VIEWPORT_NAME === 'desktop' ) {
+			promises.push(
+				this.page.waitForResponse( /async-load-calypso-blocks-jitm.*/, { timeout: 15 * 1000 } )
```
This builds a list of promises to be resolved later.

worldomonation commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -22,22 +23,12 @@ export class NavbarComponent {
 		this.page = page;
 	}
 
-	/**
-	 * Wait for load state of the page.
-	 *
-	 * @returns {Promise<void>} No return value.
-	 */
-	private async pageSettled(): Promise< void > {
-		await this.page.waitForLoadState( 'load' );
-	}
-
```
This `waitForLoadState` didn't actually do much that was useful, hence it is removed.

worldomonation committed:
packages/calypso-e2e/src/lib/components/navbar-component.ts
- apply the wait for async JITM request to all viewports

worldomonation commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -71,20 +59,17 @@ export class NavbarComponent {
 		await Promise.all( [
 			this.page.waitForLoadState( 'networkidle' ),
-			notificationsButton.waitForElementState( 'stable' ),
+			this.page.waitForResponse( /async-load-calypso-blocks-jitm.*/, { timeout: 15 * 1000 } ),
```
@WunderBart Do you think waiting on this is risky? 

I am trying to ensure that Calypso has fully loaded before attempting to interact with the NavBar > Notification button. Otherwise, in some cases the click can be swallowed up by Calypso fully loading. 

I don't really have a better request or element to key off other than this - the other, safer fallback would be `waitForLoadState(networkidle)`. What do you think?

dpasque replied:
From my perspective, I'm okay with this for now! Out of curiosity, did it not work to wait for another element on the navbar to ensure it's stable?

worldomonation replied:
@dpasque it was not sufficient to wait on other elements on the navbar. 

This can be observed by placing a breakpoint on the XHR > Load event then loading `/home`. What happens is:

The initial load occurs
![image](https://user-images.githubusercontent.com/6549265/205174718-7d2bb41f-3651-4fed-a249-4bd62bd64f61.png)

The navbar loads all at once
![image](https://user-images.githubusercontent.com/6549265/205174744-3e67c5af-0362-4a60-9a1e-60a1b6d145cc.png)

The SidebarComponent then begins the loading process
![image](https://user-images.githubusercontent.com/6549265/205174823-bee2ce8f-1ed1-44d1-9a6f-e37b5c2a54de.png)

(and then many XHR requests later)
The SidebarComponent is mostly loaded and the home page cards being loading
![image](https://user-images.githubusercontent.com/6549265/205174961-c83aed1e-d8d4-40c5-aa13-415b0b97c880.png)

Since the Navbar loads all at once, waiting on other elements contained within the Navbar is insufficient to guarantee that the page is loaded.

dpasque replied:
Makes sense -- thanks for explaining!

WunderBart replied:
(👋 Sorry for the late response; I was away for a couple of days.)

> @WunderBart Do you think waiting on this is risky?

While the above might work well as a fix for the current issue, there are two things that I'd consider risky (or even an antipattern) here:

1. Waiting for the page to be ready when initiating an element action method (scope creep).

   The scope should be focused on the element itself and the directly related events (triggered requests, animations, etc.). Waiting for the DOM events should be handled independently (if necessary); otherwise, we'd need each action method to be initiated with some kind of DOM settling waiter (as we did with `pageSettled`!), which can lead to code bloat and be easily missed if not enforced programmatically. It can also potentially lead to a situation where the same E2E Component doesn't work on another page because that page loads differently, which breaks the purpose of reusability.

2. Waiting for a response that might have already been resolved (potential false-positive).

   We should not assume that the request is pending when initiating an action method. It might currently be a case, but it could change literally any minute. It can also create a confusing debugging experience because a simple `pause()` right before the action would break the flow, as the request would most likely resolve before the method is called. In general, waiting for a response should follow an explicit request trigger, like a `Load More` button click. 
   
   ATM, I can think of two solutions here: 

   1. If the request is not directly related to the action element, it should be extracted from the action method and handled independently (e.g., at the page-load time), or
   2. Wrap the `waitForResponse` in a try/catch block so that we can continue if it times out. The long timeout value (15s) might occasionally add an extra delay, but we'd avoid a false-positive test if the request is resolved before the action is called. Still, it could be frustrating when debugging, as mentioned above. Possibly, combining it with `waitForRequest` could make it escape the try/catch block faster.
   
   My gut feeling is that it would be best to extract it from the action method and place it where the related request is being triggered. 

Hope it makes sense. What do you think, folks?

worldomonation replied:
Thanks @WunderBart, for always offering up words of wisdom.

> If the request is not directly related to the action element, it should be extracted from the action method and handled independently (e.g., at the page-load time), or

I think this will be the path I take to address this issue. The navbar should not have to be aware of the load state of the rest of the page; that's a job for either the spec to ensure, or another POM (eg. `MyHomePage`).

worldomonation replied:
Filed https://github.com/Automattic/wp-calypso/issues/70961 to make required changes.

dpasque reviewed (approved):
I think this is good to start with for now! Added one question about waiting on an element, but I think this can merge and we can watch it 👍
